### ISSUE TITLE
Google Drive spreadsheet triggers high CPU usage in Firefox (a infinite loop with browserpass?)

### ISSUE THREAD

**Issue by 4llan** [browserpass/browserpass-extension#135]
### General information

-   Operating system + version: Windows 10 1809
-   Browser + version: Firefox 66.0.3 (64-bits)
-   Information about the host app:
    -   How did you install it?
        Downloaded a pre-built binary (`browserpass-windows64-3.0.6.zip`)
-   Information about the browser extension:
    -   How did you install it?
        Webstore
    -   Browserpass extension version as reported by your browser:
        3.1.0

### Exact steps to reproduce the problem

1. Open any Google Drive Spreadsheet (https://docs.google.com/spreadsheets/) via drive.google.com

2. Firefox CPU and memory usage goes unusual high

3. `browserpass-windows64.exe` is spawned at least 50x

### What should happen?
The extension should list the available logins.

### What happened instead?
* Before opening the spreadsheet (or opening with browserpass extension disabled) ![Before opening the spreadsheet (or opening with browserpass extension disabled)](https://user-images.githubusercontent.com/2865381/56688485-42f25b00-66af-11e9-8ded-423e19c4385e.png)
* Opening spreadsheet via Google Drive ![Opening spreadsheet via Google Drive](https://user-images.githubusercontent.com/2865381/56688521-58678500-66af-11e9-9271-df22fa2f2b14.png)
* Firefox main proccess high CPU and memory usage, plus the spawning of browserpass ![Firefox main proccess high CPU and memory usage, plus the spawning of browserpass](https://user-images.githubusercontent.com/2865381/56688550-6d441880-66af-11e9-87d5-77c45dc61eb1.png)
* 5min later, the proccesses still running and `firefox.exe` still with high CPU consume ![5min later, the proccesses still running and firefox.exe still with high CPU consume](https://user-images.githubusercontent.com/2865381/56688618-94024f00-66af-11e9-998a-214446a70921.png)
* At least 50 child proccesses ![At least 50 child proccesses](https://user-images.githubusercontent.com/2865381/56688701-bb591c00-66af-11e9-83a2-0e0cef4d1d18.png)
* The badge counts the number of logins available for the domain (google.com) but don't load them ![The badge counts the number of logins available for the domain (google.com) but don't load them](https://user-images.githubusercontent.com/2865381/56688729-cad86500-66af-11e9-898d-2a88d4190c67.png)

I've noticed the high CPU usage and breakdowns after using the new browserpass. Just today I reproduce this issue by opening spreadsheets and monitoring via Process Explorer.

**Comment by maximbaz** [comment]
This is very interesting bug to investigate! 

To confirm your repro steps, do you actually open Browserpass popup at any point of time, or this all is reproducible even with the badge counter?

**Comment by 4llan** [comment]
I clicked on the popup just to screenshot and see if something changed in the child proccesses. Only with the badge counter this happens

**Comment by maximbaz** [comment]
I can't easily reproduce this on Linux, I'll try to get access to a Windows box later tonight.

In the meantime, could you please confirm that the bug reproduces in (1) private session in Firefox and (2) on [this particular spreadsheet](https://docs.google.com/spreadsheets/d/19W-lXWS9jYwqCK-LwgYo31GucPPxYVld_hVEcfpNpXg/edit#gid=1724051764) (it's just a random public one I found, want to make sure it's not caused by a specific spreadsheet).

If you can, could you please try in Chromium / Google Chrome, also in Incognito session? 

All of this would help me narrowing down the possibilities of where the bug might be.

**Comment by 4llan** [comment]
Just opened Firefox private session and the issue is reproducible via your linked spreadsheet.
![Anotação 2019-04-24 170905](https://user-images.githubusercontent.com/2865381/56690425-cada6400-66b3-11e9-9c3f-4ed9a87e7869.png)

Sorry, I can't test right now in Chromium / Chrome.

Interesting fact: after disabling the extension, the spawned proccesses are gone. But the CPU usage of `firefox.exe` continues.
![Anotação 2019-04-24 171535](https://user-images.githubusercontent.com/2865381/56690845-d1b5a680-66b4-11e9-9188-fafa161e1cd5.png)

**Comment by maximbaz** [comment]
@4llan thanks a lot! I was investigating in the meantime, and I think I found something that would at least greatly reduce the number of background processes, would you be able to test the experience in this build? Is it the same, slightly better or much better?

[browserpass-firefox-HEAD.zip](https://github.com/browserpass/browserpass-extension/files/3114022/browserpass-firefox-HEAD.zip)

**Comment by 4llan** [comment]
I loaded this temp extension and the experience is the same, @maximbaz

**Comment by 4llan** [comment]
This issue is specific from Windows. I'm using Arch Linux in another
computer with Firefox sync activated, browserpass installed via official
packages and the issue isn't reproducible here.

**Comment by 4llan** [comment]
Just a checklist to see if I can help investigating this:
- [ ] Try to track down where, in the Google Drive Spreadsheet js scripts, the tab's updated more than 50 times redundantly (status: "complete")
- [ ] Check ProcessMonitor to understand why `firefox.exe` hangs, while in linux not

**Comment by maximbaz** [comment]
With `3.1.1` out of the door, the first item is no longer relevant - regardless of the number of "status: complete" events, we now refresh counter only once the URL actually changes.

Even though you tested and it seems not to be the root cause of the issue, would be nice if you repeat the test once `3.1.1` reaches your Firefox on Windows, just to make sure 🙂

**Comment by maximbaz** [comment]
I was able to make a quick test on someone else's Chrome on Windows, 3.1.0 indeed spawned lots of processes and 3.1.1 did not, however there was no CPU spike in both cases.

Let me know if you are able to reproduce this with 3.1.1, I'm reasonably sure we have fixed the bug.

**Comment by 4llan** [comment]
Hey @maximbaz, I've tested the v3.1.1 and here's the thing:

- Chrome for Windows works as expected: no issues, no multi-spawning processes, no CPU spikes
- the issue persists in Firefox

Your change makes sense, I believe that this improved performance. However, in Firefox (both Windows and Linux) the Google Drive Spreadsheet triggers more than 50 times the tab update with the same URL. In the first update, the status is "loading"; the rest is "complete", with the same URL.

**Comment by maximbaz** [comment]
Huh, I confirm, nice catch! Well okay then, I wanted to rely on browsers to tell me when URL changes, but I guess I have no other option but to implement this myself in the extension directly. Thanks, stay tuned for 3.1.2 😉

**Comment by 4llan** [comment]
Thank you, you @maximbaz.

I was wondering: this is a bug where Firefox receive a tab update even with XHR requests?

**Comment by maximbaz** [comment]
Probably not, given that Chromium also receives a bunch of tab update events, it's just that Chromium doesn't include "url" field when it didn't change, while Firefox includes it every single time.

**Comment by ninegua** [comment]
Just want to upvote this bug report as it practically prevents the firefox extension from being used on Windows.

**Comment by maximbaz** [comment]
Oh thanks for the ping, sorry this has completely slipped my attention. Will try to look into it now.

**Comment by maximbaz** [comment]
I believe this has been fixed in `3.2.2`, please do let me know if you still experience the issue.
